[DP-89] As a learner, I want to be able to print a lesson

chanmioh opened the pull request:
### Changelog
Created a stylesheet called `print.css` which formats a lesson page. Custom classes/styles were added. The wordpress code needed to be modified for the styling to work (to create separation of elements when formatting) so for now I've modified the lesson three/Communications to work with this code: https://wp.dailp.northeastern.edu/wp-admin/post.php?post=673&action=edit

NOTE: I wasn't able to add the WP bracket encodings yet since they don't seem to work on my local environment, so it wasn't possible to test for me.

**Created CSS Classes**
`section`: Represents an element that should not be broken up between pages. (i.e. sections with headings)

<img width="219" alt="image" src="https://user-images.githubusercontent.com/40529597/192384702-86f1b3b1-2852-463d-8232-66a0ae0cbdda.png">

<img width="647" alt="image" src="https://user-images.githubusercontent.com/40529597/192381800-880d33c9-8fb9-4c90-8bbb-e1c0671e324d.png">

`prose`: Represents an element with some kind of content that needs the full-width of the page.

<img width="552" alt="image" src="https://user-images.githubusercontent.com/40529597/192381737-64de9bc2-7296-4192-807c-4d22ed11bcc5.png">


Linked file to view the formatted print pdf:
[DAILP.pdf](https://github.com/NEU-DSG/dailp-encoding/files/9649791/DAILP.pdf)

GracefulLemming reviewed (changes requested):
The output looks solid! One (admittedly large) question: Is there a reason you did not make this a typescript stylesheet (*.css.ts)? It would be more consistent with the rest of our styles at the very least.

loafofpiecrust commented:
The styles should definitely be `.css.ts` instead of plain CSS to keep consistent with the rest of our styles. We have some global rules already which you can model off of, and all of the pseudo selectors you're using should be usable there too.

loafofpiecrust commented:
@chanmioh Could you give more detail on changes you made on the WP page?

loafofpiecrust commented:
Perhaps the input boxes should have a black outline rather than gray, since gray doesn't always print well. The crafted print PDFs seem to have black borders for those.

chanmioh commented:
### Changelog
Created a `print.css.ts` file that contains the previous `print.css` styles in TypeScript. 

`print.css.ts`: Uses `globalStyles` to style elements defined under an element with the class `wp`.
`wordpress.tsx`: `parseOptions` looks for an element with a class called `print-lesson` (which is defined in the Wordpress page code) to style it with the vanilla-extract generated CSS class `wp`.
`global.css.ts`: Added some global print styles to certain elements since I assumed they should be the same for all printed pages.
     - Buttons (and input type with radio, since I assume there's no use for that in a print) are hidden.
     - Inputs are styled as a single bottom black line.


[DAILP.pdf](https://github.com/NEU-DSG/dailp-encoding/files/9678102/DAILP.pdf)

chanmioh commented:
> @chanmioh Could you give more detail on changes you made on the WP page?

Yep!

### Wordpress Changes
All changes can be viewed in: https://wp.dailp.northeastern.edu/wp-admin/post.php?post=673&action=edit

- `print-lesson`: This class was added to the first div element to distinguish WP pages that need to be printed with the lesson stylesheet.
- `section`: This class prevents elements from being broken up between pages. This was used for creating pagebreaks for each new `h1` (i.e. 1. Introduction, 2. Reading ...) and for keeping sections of the lesson page together such as related lists of questions, or vocabulary. 
<img width="360" alt="image" src="https://user-images.githubusercontent.com/40529597/193115231-afe6e1e3-e12b-4a1e-a8e3-4812d4a5cfdd.png">
- `prose`: Some content requires being full-width such as the introductory prose, and I saw this content was being broken up by `<br/>`. This class styles certain elements to hide this, since some elements require it for formatting purposes.
<img width="340" alt="image" src="https://user-images.githubusercontent.com/40529597/193115463-4926bc09-7d09-4acd-b220-bee0879ef52c.png">
- `syllabary-input`: Some input fields should be numbered with large textboxes rather than just the normal black lines, so I created this class to account for those fields. There was also a `show-on-print` class included in this formatting since the numbered textboxes shouldn't show on the website, but on the print.
<img width="326" alt="image" src="https://user-images.githubusercontent.com/40529597/193115692-7d0a709d-424a-4b7f-bdf0-3b9b8b60a1c9.png">

<img width="437" alt="image" src="https://user-images.githubusercontent.com/40529597/193116446-2f7107f7-0397-454c-9ab5-016d62a3a583.png">

- `short-answer`: This was styled so the input fields take up the rest of the space of a line with the question, rather than starting on the next line.
<img width="336" alt="image" src="https://user-images.githubusercontent.com/40529597/193116001-ffc3a23a-3fef-41d7-9835-59e23ffd9539.png">

<img width="437" alt="image" src="https://user-images.githubusercontent.com/40529597/193116513-8513175d-a140-4d2b-8db4-a13c62221619.png">

- `centered`: A class used to style the rows of a `table` element, since I needed the image and its blue text centered in a row.
<img width="346" alt="image" src="https://user-images.githubusercontent.com/40529597/193116405-ecc7b909-5f05-428f-ad75-8b31dd1d45b8.png">

<img width="452" alt="image" src="https://user-images.githubusercontent.com/40529597/193116544-b0d252c8-575e-4fa1-9c39-c7c490f1040a.png">

loafofpiecrust commented:
> wordpress.tsx: parseOptions looks for an element with a class called print-lesson (which is defined in the Wordpress page code) to style it with the vanilla-extract generated CSS class wp.

Would it make sense to instead check the URL we're rendering or apply to all WP pages? It's okay if you think this is the best solution, I'm just trying to reduce the amount of special handling we need to do on the WP end.

loafofpiecrust commented:
> * `prose`: Some content requires being full-width such as the introductory prose, and I saw this content was being broken up by `<br/>`. This class styles only certain elements to hide this, since some elements require it for formatting purposes.

Oh! This one is weird. Ideally, the WP paragraphs shouldn't have `<br/>` tags inside of it and should flow with the page. This should be an editing artifact you can get rid of by removing newlines in the WP page editor. Does that sound about right to you?

loafofpiecrust commented on website/src/components/print-lesson.css.ts:
```diff
@@ -0,0 +1,144 @@
+const section = `${wp} .section`
+
+// This removes line breaks to create full-width pages.
+globalStyle(`${prose} br`, {
+  ...media(mediaQueries.print, {
```
Does the `globalStyle` function stop you from doing `globalStyle(..., media(mediaQueries.print, ...))` directly?

chanmioh replied:
yeah, I have to do `...media()` anytime I want to use that function. since the `...media()` makes the code look a bit cluttered, I can create a function within this file that does the `...media()` call which the globalStyles will call instead, which may look cleaner.

chanmioh replied:
oh actually I can do `globalStyle(..., media(mediaQueries.print, ...))` just fine, I just had to get rid of the `{}` between `...media()`. I can still go ahead and just create a function since there's a lot of duplicate code involving `mediaQueries.print`

loafofpiecrust commented on website/src/components/wordpress.tsx:
```diff
@@ -78,6 +81,12 @@ const parseOptions: HTMLReactParserOptions = {
         return <Link {...props}>{domToReact(node.children, parseOptions)}</Link>
+      } else if (node.attribs["class"]?.includes("print-lesson")) {
+        return (
+          <div className={wpCss.wp}>
```
this css style class probably deserves a more specific name, since it's not for just any WP content so far.

chanmioh replied:
sounds good, renaming this to `lesson`!

chanmioh commented:
> > * `section`: This class prevents elements from being broken up between pages. This was used for creating pagebreaks for each new section of content (i.e. 1. Introduction, 2. Reading ...) and for keeping sections of the lesson page together such as related lists of questions, or vocabulary.
> 
> Hmmm, it looks like all of the top-level elements have `section`, is that right? If so, I think you can do `.print-lesson > *` (all direct descendants) instead of `.print-lesson .section` (all `.section` descendants recursive) for the selector.

The `section` was also used for some descendants of top-level elements with `section`, here's an example: 

<img width="248" alt="image" src="https://user-images.githubusercontent.com/40529597/193328489-89ce4e2c-7afa-4e26-a49e-5f5061915425.png">

This allows for keeping together parts of the lesson that aren't at the top-level, but still needs to start on a new page if not all of its content can fit on one page.

chanmioh commented:
> > * `prose`: Some content requires being full-width such as the introductory prose, and I saw this content was being broken up by `<br/>`. This class styles only certain elements to hide this, since some elements require it for formatting purposes.
> 
> Oh! This one is weird. Ideally, the WP paragraphs shouldn't have `<br/>` tags inside of it and should flow with the page on screen and print. This should be an editing artifact you can get rid of by removing newlines in the WP page editor. Does that sound about right to you?

Gotcha, I was wondering if the `<br/>` was intentional, so I didn't want to mess with that in the WP page. I can go ahead and remove those new lines though and get rid of this CSS class!

loafofpiecrust commented:
> Gotcha, I was wondering if the `<br/>` was intentional, so I didn't want to mess with that in the WP page.

It's hard to get nice HTML using the visual editor in WP 😦

loafofpiecrust commented:
@chanmioh If any of these suggestions seem like a bigger task, we can put them off for now. Just say the word.

chanmioh commented:
> > * `centered`: A class used to style the rows of a `table` element, since I needed the image and its blue text centered in a row.
> 
> Hmm these aren't semantic tables, so they shouldn't really be table elements ideally. They should be [`fig*` elements](https://www.w3schools.com/tags/tag_figcaption.asp), perhaps if we use those elements that can replace the css class?

Yeah that makes sense - I'll go ahead and change those to be figure elements!

chanmioh commented:
### Changelog

- `WP page`:
     - Removed `prose` element.
     - Removed unnecessary linebreaks between sentences that should appear both on-screen and print media.
     - Replaced `<table>` element when utilized for image and caption pairs with a `<figure>` element. Styled the figure element with flexbox accordingly. It should look very similar to how it did as a table:
<img width="1022" alt="image" src="https://user-images.githubusercontent.com/40529597/193358913-37adad21-62ce-4b97-b07c-01b27072a983.png">

     - Added a `<hr>` element underneath the `<figure>` element
<img width="1011" alt="image" src="https://user-images.githubusercontent.com/40529597/193358958-fa698fa7-bfd3-4a05-8ba2-a8977ee44af8.png">

- `global.css.ts`: `<hr>` styled to not appear on print.
- `wordpress.tsx`: Checked for URL path before parsing to see if the new print stylesheet needs to be included.
- `print-lesson.css.ts`: 
     - Added function `printMedia` to minimize duplicate code of specifying the media query.
     - Removed styles for `<table>`.
     - Added styled for `<figure>`.

GracefulLemming reviewed (approved):
With the changes made, this looks really nice! @chanmioh, we should follow-up about documenting the WP changes in the wiki so our content team can reference it!

chanmioh commented:
> With the changes made, this looks really nice! @chanmioh, we should follow-up about documenting the WP changes in the wiki so our content team can reference it!

Sounds good! I'll go ahead and merge this in for now.
